### [Problem 1] Execution of various methods

## LSTM

In [1]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing import sequence

# Parameters
max_features = 20000
maxlen = 80
batch_size = 32
epochs = 5

# Load data
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)

# Pad sequences
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)

# Model
model = Sequential()
model.add(Embedding(max_features, 128))
model.add(LSTM(128, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(1, activation='sigmoid'))

# Compile model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# Train model
model.fit(x_train, y_train, batch_size=batch_size, epochs=epochs, validation_data=(x_test, y_test))

# Evaluate model
score, acc = model.evaluate(x_test, y_test, batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)


2024-07-31 12:14:55.593997: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-07-31 12:14:56.869536: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


Epoch 1/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 114s 142ms/step - accuracy: 0.7150 - loss: 0.5418 - val_accuracy: 0.8344 - val_loss: 0.3789
Epoch 2/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 98s 125ms/step - accuracy: 0.8762 - loss: 0.3046 - val_accuracy: 0.8300 - val_loss: 0.3847
Epoch 3/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 99s 126ms/step - accuracy: 0.9131 - loss: 0.2238 - val_accuracy: 0.8330 - val_loss: 0.3956
Epoch 4/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 100s 127ms/step - accuracy: 0.9433 - loss: 0.1585 - val_accuracy: 0.8300 - val_loss: 0.4672
Epoch 5/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 96s 123ms/step - accuracy: 0.9598 - loss: 0.1099 - val_accuracy: 0.8210 - val_loss: 0.5174
782/782 ━━━━━━━━━━━━━━━━━━━━ 20s 25ms/step - accuracy: 0.8192 - loss: 0.5264
Test score: 0.5173606276512146
Test accuracy: 0.8209999799728394


## GRU

In [2]:
from tensorflow.keras.layers import GRU

model = Sequential()
model.add(Embedding(max_features, 128))
model.add(GRU(128, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(1, activation='sigmoid'))

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(x_train, y_train, batch_size=batch_size, epochs=epochs, validation_data=(x_test, y_test))

score, acc = model.evaluate(x_test, y_test, batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)


Epoch 1/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 88s 109ms/step - accuracy: 0.6380 - loss: 1.2565 - val_accuracy: 0.6790 - val_loss: 0.5877
Epoch 2/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 80s 102ms/step - accuracy: 0.7665 - loss: 0.4925 - val_accuracy: 0.6972 - val_loss: 0.5722
Epoch 3/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 85s 108ms/step - accuracy: 0.8295 - loss: 122.9253 - val_accuracy: 0.6336 - val_loss: 0.6432
Epoch 4/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 84s 107ms/step - accuracy: 0.7498 - loss: 0.5372 - val_accuracy: 0.6578 - val_loss: 0.6302
Epoch 5/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 83s 106ms/step - accuracy: 0.8183 - loss: 0.4102 - val_accuracy: 0.6632 - val_loss: 0.6336
782/782 ━━━━━━━━━━━━━━━━━━━━ 14s 18ms/step - accuracy: 0.6592 - loss: 0.6407
Test score: 0.6336373686790466
Test accuracy: 0.6632000207901001


## SimpleRNN

In [3]:
from tensorflow.keras.layers import SimpleRNN

model = Sequential()
model.add(Embedding(max_features, 128))
model.add(SimpleRNN(128, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(1, activation='sigmoid'))

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(x_train, y_train, batch_size=batch_size, epochs=epochs, validation_data=(x_test, y_test))

score, acc = model.evaluate(x_test, y_test, batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)


Epoch 1/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 48s 59ms/step - accuracy: 0.5109 - loss: 0.7057 - val_accuracy: 0.6059 - val_loss: 0.6659
Epoch 2/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 45s 58ms/step - accuracy: 0.6433 - loss: 0.6268 - val_accuracy: 0.6267 - val_loss: 0.6319
Epoch 3/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 48s 61ms/step - accuracy: 0.7305 - loss: 0.5305 - val_accuracy: 0.6974 - val_loss: 0.5886
Epoch 4/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 47s 60ms/step - accuracy: 0.8062 - loss: 0.4217 - val_accuracy: 0.7114 - val_loss: 0.5981
Epoch 5/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 47s 61ms/step - accuracy: 0.8396 - loss: 0.3669 - val_accuracy: 0.7454 - val_loss: 0.6079
782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 7ms/step - accuracy: 0.7413 - loss: 0.6145
Test score: 0.6078584790229797
Test accuracy: 0.7453600168228149


The results of running the different RNN models (LSTM, GRU, SimpleRNN) on the IMDB dataset are as follows:

### LSTM:

Test score: 0.5174

Test accuracy: 82.1%

### GRU:

Test score: 0.6336

Test accuracy: 66.3%

### SimpleRNN:

Test score: 0.6079

Test accuracy: 74.5%

### Analysis
#### LSTM (Long Short-Term Memory)

The LSTM model outperforms both GRU and SimpleRNN in terms of test accuracy, achieving an accuracy of 82.1%.
LSTM's ability to capture long-term dependencies makes it particularly effective for the IMDB sentiment analysis task.

#### GRU (Gated Recurrent Unit)

The GRU model shows a lower performance compared to LSTM, with a test accuracy of 66.3%.
While GRUs are also designed to handle long-term dependencies, they are sometimes less powerful than LSTMs for certain tasks, though they can be more computationally efficient.

#### SimpleRNN

The SimpleRNN model achieves a test accuracy of 74.5%, which is better than GRU but not as good as LSTM.
SimpleRNNs struggle with capturing long-term dependencies, which might explain why their performance is not as strong as LSTM's.

#### Conclusion
From the above results, it is evident that LSTM performs the best for the IMDB sentiment analysis task. This reinforces the understanding that LSTM networks are well-suited for tasks requiring the modeling of long-term dependencies in sequences.

### [Problem 2] (Advance assignment) Comparison between multiple data sets

In [4]:
from tensorflow.keras.datasets import reuters
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.layers import Dropout

max_words = 10000
batch_size = 32

# Load data
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=max_words)

# Pad sequences
tokenizer = Tokenizer(num_words=max_words)
x_train = tokenizer.sequences_to_matrix(x_train, mode='binary')
x_test = tokenizer.sequences_to_matrix(x_test, mode='binary')

# Convert labels to categorical
num_classes = np.max(y_train) + 1
y_train = tf.keras.utils.to_categorical(y_train, num_classes)
y_test = tf.keras.utils.to_categorical(y_test, num_classes)

# Model
model = Sequential()
model.add(Dense(512, input_shape=(max_words,), activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(x_train, y_train, batch_size=batch_size, epochs=epochs, validation_data=(x_test, y_test))

score, acc = model.evaluate(x_test, y_test, batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)


/home/topister/Desktop/dpro/machine-learning/dpro/lib/python3.10/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
2024-07-31 12:35:08.995581: W external/local_tsl/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 359280000 exceeds 10% of free system memory.


Epoch 1/5
281/281 ━━━━━━━━━━━━━━━━━━━━ 0s 84ms/step - accuracy: 0.6245 - loss: 1.7499

2024-07-31 12:42:52.557090: W external/local_tsl/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 89840000 exceeds 10% of free system memory.


281/281 ━━━━━━━━━━━━━━━━━━━━ 27s 88ms/step - accuracy: 0.6249 - loss: 1.7481 - val_accuracy: 0.8028 - val_loss: 0.8690
Epoch 2/5
281/281 ━━━━━━━━━━━━━━━━━━━━ 24s 85ms/step - accuracy: 0.8958 - loss: 0.4847 - val_accuracy: 0.8090 - val_loss: 0.8312
Epoch 3/5
281/281 ━━━━━━━━━━━━━━━━━━━━ 24s 84ms/step - accuracy: 0.9366 - loss: 0.2762 - val_accuracy: 0.8103 - val_loss: 0.8739
Epoch 4/5
281/281 ━━━━━━━━━━━━━━━━━━━━ 23s 83ms/step - accuracy: 0.9554 - loss: 0.1970 - val_accuracy: 0.8117 - val_loss: 0.8999
Epoch 5/5
281/281 ━━━━━━━━━━━━━━━━━━━━ 23s 82ms/step - accuracy: 0.9499 - loss: 0.2015 - val_accuracy: 0.8108 - val_loss: 0.9298
16/71 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8200 - loss: 0.9200

2024-07-31 12:44:27.596081: W external/local_tsl/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 89840000 exceeds 10% of free system memory.


71/71 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.8168 - loss: 0.8992
Test score: 0.9298367500305176
Test accuracy: 0.8107746839523315


### [Problem 3] Explanation of other classes

RNN - The base class for all recurrent layers.

SimpleRNNCell - The cell class for SimpleRNN, representing the internal computations of SimpleRNN.

GRUCell - The cell class for GRU, representing the internal computations of GRU.

LSTMCell - The cell class for LSTM, representing the internal computations of LSTM.

StackedRNNCells - A wrapper for stacking multiple RNN cells.

CuDNNGRU - A fast GRU implementation leveraging NVIDIA's CuDNN library for GPU acceleration.

CuDNNLSTM - A fast LSTM implementation leveraging NVIDIA's CuDNN library for GPU acceleration.